File: src/main/java/seremis/geninfusion/api/util/Line2D.java
package seremis.geninfusion.api.util;

public class Line2D {

    public Coordinate2D head, tail;

    public Line2D(Coordinate2D head, Coordinate2D tail) {
        this.head = head;
        this.tail = tail;
    }

    public Line2D(double x1, double y1, double x2, double y2) {
        this.head = new Coordinate2D(x1, y1);
        this.tail = new Coordinate2D(x2, y2);
    }

    public Line2D() {
        this(0, 0, 0, 0);
    }

    public Coordinate2D getHead() {
        return this.head;
    }

    public Coordinate2D getTail() {
        return this.tail;
    }

    public void setLine(Coordinate2D head, Coordinate2D tail) {
        this.head = head;
        this.tail = tail;
    }

    public void setLine(double x1, double y1, double x2, double y2) {
        this.head = new Coordinate2D(x1, y1);
        this.tail = new Coordinate2D(x2, y2);
    }

    public Line2D setLine(Line3D coord) {
        this.head = coord.head.to2D();
        this.tail = coord.tail.to2D();
        return this;
    }

    private double length = -1;
    
    public double getLength() {
        if(length == -1) {
            double dx = Math.abs(head.getXCoord() - tail.getXCoord());
            double dy = Math.abs(head.getYCoord() - tail.getYCoord());
            length = Math.sqrt(dx * dx + dy * dy);
        }
        return length;
    }

    private double yaw = -1;
    
    public double getYaw() {
        if(yaw == -1) {
            double xd = head.x - tail.x;
            double yd = head.y - tail.y;
    
            yaw = Math.atan2(xd, yd) * 180.0D / Math.PI;
        }
        return yaw;
    }

    @Override
    public String toString() {
        return "Line2D[head x: " + head.x + ", y: " + head.y + " tail x: " + tail.x + ", y: " + tail.y + "]";
    }
}

File: src/main/java/seremis/geninfusion/api/util/Line2D.scala
package seremis.geninfusion.api.util

class Line2D(var head: Coordinate2D, var tail: Coordinate2D) {

    def this(x1: Double, y1: Double, x2: Double, y2: Double) {
        this(new Coordinate2D(x1, y1), new Coordinate2D(x2, y2))
    }

    def getHead: Coordinate2D = this.head

    def getTail: Coordinate2D = this.tail

    def setLine(head: Coordinate2D, tail: Coordinate2D) {
        this.head = head
        this.tail = tail
    }

    def setLine(x1: Double, y1: Double, x2: Double, y2: Double) {
        this.head = new Coordinate2D(x1, y1)
        this.tail = new Coordinate2D(x2, y2)
    }

    def getLength: Double = {
        val dx = Math.abs(head.getXCoord - tail.getXCoord)
        val dy = Math.abs(head.getYCoord - tail.getYCoord)
        Math.sqrt(dx * dx + dy * dy)
    }

    def getYaw: Double = {
        val xd = head.x - tail.x
        val yd = head.y - tail.y
        Math.atan2(xd, yd) * 180.0D / Math.PI
    }

    override def toString: String = "Line2D[head x: " + head.x + ", y: " + head.y + " tail x: " + tail.x + ", y: " + tail.y + "]"

}